tiledb/stdx/synchronized_optional/controlled_reference.h

```diff
@@ -0,0 +1,255 @@
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ *
+ * @section DESCRIPTION
+ *
+ * This file implements `constrolled_reference` and `controlled_const_reference`
+ *
+ * A controlled reference provides a scope-oriented access to another object. In
+ * these cases the reference is a resource that needs to be "opened" and
+ * "closed", whatever that might mean for the resource in question.
+ *
+ * A controlled reference is a wrapper around a handle supplied by the referent
+ * class. A referent may supply more than one handle if needed. A traits
+ * template argument selects which handle is used. These classes may be
+ * considered boilerplate for a scoped access class in a way that's common to
+ * the pattern as a whole.
+ *
+ * Controlled references come in both `const` and non-`const` varieties. These
+ * may or may not use the same handle class.
+ *
+ * The handle interface is asymmetric. The referent presents an `attach`
+ * function that supplies a handle. The corresponding `detach` function is
+ * called on the handle itself.
+ */
+
+#ifndef TILEDB_CONTROLLED_REFERENCE_H
+#define TILEDB_CONTROLLED_REFERENCE_H
+
+#include "tiledb/common/common.h"
+
+namespace stdx {
+namespace detail {
+
+/**
+ * Default traits class for controlled reference classes. Both `const` and
+ * non-`const` varieties use the same traits.
+ *
+ * @tparam H The holder class that contains a referent object.
+ */
+template <class H>
+class [[maybe_unused]] controlled_reference_default_traits {
+  /**
+   * The handle type is supplied by the holder class.
+   */
+  using handle_type = typename H::handle_type;
+
+  /**
+   * Ordinary factory function for handles.
+   *
+   * @param source
+   * @return
+   */
+  handle_type attach(H& source) {
+    return source.attach();
+  }
+
+  /**
+   * Factory function for handles, `adopt_lock` variant.
+   *
+   * @param source
+   * @return
+   */
+  handle_type attach(H& source, std::adopt_lock_t) {
+    return source.attach(std::adopt_lock);
+  }
+};
+
+/**
+ * Base class for reference objects.
+ *
+ * @tparam T The type that this class is a reference to, as if it were `T&`
+ * @tparam H A class that holds an object of type `T`
+ * @tparam Tr Traits for H<T> allows different kinds of references to a holder
+ */
+template <
```

Comment:
The next two classes might benefit from extra documentation. I really appreciated the level of documentation in the synchronized_optional.h file.